staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go

```diff
@@ -158,3 +164,44 @@ func compileRule(rule apiextensions.ValidationRule, env *cel.Env) (compilationRe
 func generateUniqueSelfTypeName() string {
 	return fmt.Sprintf("selfType%d", time.Now().Nanosecond())
 }
+
+type celEstimator struct {
+	root *celmodel.DeclType
+}
+
+func (c *celEstimator) EstimateSize(element checker.AstNode) *checker.SizeEstimate {
+	if len(element.Path()) == 0 {
+		// Path() can return an empty list, early exit if it does since we can't
+		// provide size estimates when that happens
+		return nil
+	}
+	currentNode := c.root
+	// cut off "self" from path, since we always start there
+	for _, name := range element.Path()[1:] {
+		var nextNode *celmodel.DeclType
+		switch name {
+		case "@items", "@values":
+			nextNode = currentNode.ElemType
```

Comment:
I'm not understanding how this traversal works... should we be assigning `currentNode = nextNode` somewhere? nextNode is local to the loop but doesn't get used anywhere I can see... are there tests exercising the @items / @values case?

Reply:
Oops, you're right that `currentNode` wasn't getting reassigned. The `"@items"` and `"@values"` codepath get tested wherever we're concerned with the maximum length/size of an array member, so the `index of array with numbers` and `index of array with strings` cases both follow that codepath, I believe. `map of strings with value length` tests out the `@keys` codepath since it uses a map.

It might be useful to have a tool that explains exactly how and why a particular cost was estimated (the incorrect codepath resulted in estimates that were only subtly different rather than wildly different outside of the map case) - @jpbetz and I have discussed implementing that kind of thing.

Reply:
Clearly we need more tests. In this case, writing the test that breaks (and then fixing it), is a good workflow.

Reply:
The unit test suite has been significantly expanded - leaving the conversation open for the time being just in case anyone had anything to add, though.